File: app/api/v1/reminders.py
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_session
from app.core.deps import PaginationParams, get_current_user
from app.crud.reminder import (
    create_reminder,
    get_reminder,
    list_reminders,
    soft_delete_reminder,
    update_reminder,
)
from app.models.user import User
from app.schemas.paginated import Paginated
from app.schemas.reminder import ReminderCreate, ReminderPublic, ReminderUpdate

router = APIRouter(prefix="/reminders", tags=["reminders"])
person_reminders_router = APIRouter(
    prefix="/persons/{person_id}/reminders", tags=["reminders"]
)
asset_reminders_router = APIRouter(
    prefix="/assets/{asset_id}/reminders", tags=["reminders"]
)
subscription_reminders_router = APIRouter(
    prefix="/subscriptions/{sub_id}/reminders", tags=["reminders"]
)


@router.post("/", response_model=ReminderPublic, status_code=status.HTTP_201_CREATED)
async def create(
    data: ReminderCreate,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    return await create_reminder(db, current_user.id, data)


@router.get("/", response_model=Paginated[ReminderPublic])
async def list_all(
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
    pagination: Annotated[PaginationParams, Depends(PaginationParams)],
    is_done: bool | None = None,
):
    items, total = await list_reminders(
        db, current_user.id, skip=pagination.skip, limit=pagination.limit, is_done=is_done
    )
    return Paginated(items=items, total=total, skip=pagination.skip, limit=pagination.limit)


@router.get("/{reminder_id}", response_model=ReminderPublic)
async def get_one(
    reminder_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    reminder = await get_reminder(db, reminder_id, current_user.id)
    if not reminder:
        raise HTTPException(status_code=404, detail="Reminder not found")
    return reminder


@router.patch("/{reminder_id}", response_model=ReminderPublic)
async def patch(
    reminder_id: uuid.UUID,
    data: ReminderUpdate,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    reminder = await update_reminder(db, reminder_id, current_user.id, data)
    if not reminder:
        raise HTTPException(status_code=404, detail="Reminder not found")
    return reminder


@router.delete("/{reminder_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete(
    reminder_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    deleted = await soft_delete_reminder(db, reminder_id, current_user.id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Reminder not found")


# ── Entity-scoped reminders ──────────────────────────────────────────────────


@person_reminders_router.get("/", response_model=Paginated[ReminderPublic])
async def list_person_reminders(
    person_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
    pagination: Annotated[PaginationParams, Depends(PaginationParams)],
):
    items, total = await list_reminders(
        db, current_user.id, skip=pagination.skip, limit=pagination.limit, person_id=person_id
    )
    return Paginated(items=items, total=total, skip=pagination.skip, limit=pagination.limit)


@asset_reminders_router.get("/", response_model=Paginated[ReminderPublic])
async def list_asset_reminders(
    asset_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
    pagination: Annotated[PaginationParams, Depends(PaginationParams)],
):
    items, total = await list_reminders(
        db, current_user.id, skip=pagination.skip, limit=pagination.limit, asset_id=asset_id
    )
    return Paginated(items=items, total=total, skip=pagination.skip, limit=pagination.limit)


@subscription_reminders_router.get("/", response_model=Paginated[ReminderPublic])
async def list_subscription_reminders(
    sub_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
    pagination: Annotated[PaginationParams, Depends(PaginationParams)],
):
    items, total = await list_reminders(
        db, current_user.id, skip=pagination.skip, limit=pagination.limit, subscription_id=sub_id
    )
    return Paginated(items=items, total=total, skip=pagination.skip, limit=pagination.limit)

File: frontend/src/api/reminders.ts
import { api } from "./client";
import type { Paginated } from "./types";

export interface ReminderPublic {
  id: string;
  owner_id: string;
  title: string;
  body: string | null;
  due_at: string;
  remind_at: string | null;
  recurrence: string | null;
  is_done: boolean;
  done_at: string | null;
  person_id: string | null;
  asset_id: string | null;
  subscription_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface ReminderCreatePayload {
  title: string;
  body?: string | null;
  due_at: string;
  remind_at?: string | null;
  recurrence?: string | null;
  person_id?: string | null;
  asset_id?: string | null;
  subscription_id?: string | null;
}

export interface ReminderUpdatePayload {
  title?: string;
  body?: string | null;
  due_at?: string;
  remind_at?: string | null;
  recurrence?: string | null;
  is_done?: boolean;
}

export function listReminders(params: {
  skip?: number;
  limit?: number;
  is_done?: boolean;
} = {}): Promise<Paginated<ReminderPublic>> {
  const q = new URLSearchParams();
  if (params.skip !== undefined) q.set("skip", String(params.skip));
  if (params.limit !== undefined) q.set("limit", String(params.limit));
  if (params.is_done !== undefined) q.set("is_done", String(params.is_done));
  const qs = q.toString();
  return api.get<Paginated<ReminderPublic>>(`/reminders${qs ? `?${qs}` : ""}`);
}

export function getReminder(id: string): Promise<ReminderPublic> {
  return api.get<ReminderPublic>(`/reminders/${id}`);
}

export function createReminder(data: ReminderCreatePayload): Promise<ReminderPublic> {
  return api.post<ReminderPublic>("/reminders", data);
}

export function updateReminder(id: string, data: ReminderUpdatePayload): Promise<ReminderPublic> {
  return api.patch<ReminderPublic>(`/reminders/${id}`, data);
}

export function deleteReminder(id: string): Promise<void> {
  return api.delete(`/reminders/${id}`);
}
